plugins/GSdx/res/tfx.fx

```diff
@@ -632,37 +625,145 @@ float4 ps_color(PS_INPUT input)
 #endif
 
 #if PS_CHANNEL_FETCH == 1
-	float4 t = fetch_red(int2(input.p.xy));
+	float4 T = fetch_red(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 2
-	float4 t = fetch_green(int2(input.p.xy));
+	float4 T = fetch_green(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 3
-	float4 t = fetch_blue(int2(input.p.xy));
+	float4 T = fetch_blue(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 4
-	float4 t = fetch_alpha(int2(input.p.xy));
+	float4 T = fetch_alpha(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 5
-	float4 t = fetch_rgb(int2(input.p.xy));
+	float4 T = fetch_rgb(int2(input.p.xy));
 #elif PS_CHANNEL_FETCH == 6
-	float4 t = fetch_gXbY(int2(input.p.xy));
+	float4 T = fetch_gXbY(int2(input.p.xy));
 #elif PS_DEPTH_FMT > 0
-	float4 t = sample_depth(st_int, input.p.xy);
+	float4 T = sample_depth(st_int, input.p.xy);
 #else
-	float4 t = sample_color(st);
+#define CHUCHU
+	float4 T = sample_color(st);
 #endif
 
-	float4 c = tfx(t, input.c);
+	float4 xC = trunc(input.c * 255.0f + 0.5f); // FIX THIS
+	float4 C = tfx(T, xC);
+	
+	atst(C);
 
-	atst(c);
+	C = fog(C, input.t.z);
+	
+	if(PS_CLR1) // needed for Cd * (As/Ad/F + 1) blending modes
+	{
+		C.rgb = (float3)255.0f;
+	}
 
-	c = fog(c, input.t.z);
+	return C;
+}
 
-	if(PS_CLR1) // needed for Cd * (As/Ad/F + 1) blending modes
+void ps_fbmask(inout float4 C, float2 pos_xy)
+{
+	if (PS_FBMASK)
 	{
-		c.rgb = 1;
+		float4 rt = RtSampler.Load(int3(pos_xy, 0));
+		uint4 denorm_rt = uint4(rt * 255.0f + 0.1f);
+		uint4 denorm_c = uint4(C);
+		C = float4((denorm_c & ~FbMask) | (denorm_rt & FbMask)) / 255.0f;
```

Comment:
/ 255.0f not needed?

Reply:
Actually you can use gl code here too for the entire function.

Reply:
Nope, not needed. Check the new code please.

Reply:
Seems good now.